File: C++/chapter3/duplicateChecker.cpp
#include <iostream>
#include <vector>
#include <cassert>
using namespace std;

class Solution {
public:
    int removeDuplicates(vector<int>& nums) {
        if(nums.empty()) return 0;
        
        int a = 1;
        int cm = 1;
        int res = 1;
        int n = nums.size();

        while(cm < n){
            if(nums[cm] == nums[cm-1]){
                cm++;
                continue;
            }
            nums[a] = nums[cm];
            a++;
            cm++;
            res++;
        }
        return res;
    }
};

// Helper function to print array
void printArray(vector<int>& nums, int length) {
    cout << "[";
    for(int i = 0; i < length; i++) {
        cout << nums[i];
        if(i < length - 1) cout << ",";
    }
    cout << "]" << endl;
}

int main() {
    Solution solution;
    
    // Test Case 1: Basic case with duplicates
    cout << "Test Case 1: [1,1,2]" << endl;
    vector<int> nums1 = {1,1,2};
    int result1 = solution.removeDuplicates(nums1);
    cout << "Length: " << result1 << ", Array: ";
    printArray(nums1, result1);
    assert(result1 == 2);
    cout << "PASSED\n" << endl;
    
    // Test Case 2: More duplicates
    cout << "Test Case 2: [0,0,1,1,1,2,2,3,3,4]" << endl;
    vector<int> nums2 = {0,0,1,1,1,2,2,3,3,4};
    int result2 = solution.removeDuplicates(nums2);
    cout << "Length: " << result2 << ", Array: ";
    printArray(nums2, result2);
    assert(result2 == 5);
    cout << "PASSED\n" << endl;
    
    // Test Case 3: No duplicates
    cout << "Test Case 3: [1,2,3,4,5]" << endl;
    vector<int> nums3 = {1,2,3,4,5};
    int result3 = solution.removeDuplicates(nums3);
    cout << "Length: " << result3 << ", Array: ";
    printArray(nums3, result3);
    assert(result3 == 5);
    cout << "PASSED\n" << endl;
    
    // Test Case 4: All same elements
    cout << "Test Case 4: [1,1,1,1]" << endl;
    vector<int> nums4 = {1,1,1,1};
    int result4 = solution.removeDuplicates(nums4);
    cout << "Length: " << result4 << ", Array: ";
    printArray(nums4, result4);
    assert(result4 == 1);
    cout << "PASSED\n" << endl;
    
    // Test Case 5: Single element
    cout << "Test Case 5: [1]" << endl;
    vector<int> nums5 = {1};
    int result5 = solution.removeDuplicates(nums5);
    cout << "Length: " << result5 << ", Array: ";
    printArray(nums5, result5);
    assert(result5 == 1);
    cout << "PASSED\n" << endl;
    
    // Test Case 6: Two identical elements
    cout << "Test Case 6: [1,1]" << endl;
    vector<int> nums6 = {1,1};
    int result6 = solution.removeDuplicates(nums6);
    cout << "Length: " << result6 << ", Array: ";
    printArray(nums6, result6);
    assert(result6 == 1);
    cout << "PASSED\n" << endl;
    
    // Test Case 7: Negative numbers
    cout << "Test Case 7: [-3,-1,-1,0,0,0,0,1,1]" << endl;
    vector<int> nums7 = {-3,-1,-1,0,0,0,0,1,1};
    int result7 = solution.removeDuplicates(nums7);
    cout << "Length: " << result7 << ", Array: ";
    printArray(nums7, result7);
    assert(result7 == 4);
    cout << "PASSED\n" << endl;
    
    cout << "All test cases PASSED! ✓" << endl;
    return 0;
}

// class Solution {
// public:
//     int removeDuplicates(vector<int>& nums) {
//         int a=1;
//         int cm=1;
//         int res=1;
//         int n=nums.size();

//         while(cm<n){
//             if(nums[cm] == nums[cm-1]){
//                 cm++;
//                 continue;
//             }
//             nums[a]=nums[cm];
//             a++;
//             cm++;
//             res++;
//         }
//         return res;

        
//     }
// };

// for my reference 

File: Python/chap3/duplicateChecker.py
def removeDuplicates(nums):
    """
    Remove duplicates from sorted array in-place
    Returns the number of unique elements
    """
    if not nums:
        return 0
    
    a = 0       # Write position
    unique = 1  # Count of unique elements (first element always unique)
    count = 1   # Read position (start from index 1)
    n = len(nums)
    
    while count < n:
        # If current element is same as previous, skip it
        if count > 0 and nums[count] == nums[count-1]:
            count += 1
            continue
        
        # Found a unique element, place it at write position
        a += 1
        nums[a] = nums[count]
        unique += 1
        count += 1
    
    return unique

# Test the function
nums = [0,0,1,1,1,2,2,3,3,4]
print(f"Original array: {nums}")

unique_count = removeDuplicates(nums)

print(f"Array after removing duplicates: {nums[:unique_count]} + {['_'] * (len(nums) - unique_count)}")
print(f"Number of unique elements: {unique_count}")

# Alternative: More Pythonic solutions
def removeDuplicates_pythonic1(nums):
    """Using set to remove duplicates (but loses order)"""
    return len(set(nums))

def removeDuplicates_pythonic2(nums):
    """Using list comprehension with enumerate"""
    if not nums:
        return 0
    
    write_pos = 1
    for i in range(1, len(nums)):
        if nums[i] != nums[i-1]:
            nums[write_pos] = nums[i]
            write_pos += 1
    
    return write_pos

def removeDuplicates_pythonic3(nums):
    """Using two pointers - most readable"""
    if not nums:
        return 0
    
    slow = 0
    for fast in range(1, len(nums)):
        if nums[fast] != nums[slow]:
            slow += 1
            nums[slow] = nums[fast]
    
    return slow + 1

# Test all methods
print("\n=== Different Python Solutions ===")
test_nums = [0,0,1,1,1,2,2,3,3,4]

# Method 1 (Your approach - fixed)
nums1 = test_nums.copy()
result1 = removeDuplicates(nums1)
print(f"Method 1 (Your approach): {result1} unique elements")

# Method 2 (List comprehension)
nums2 = test_nums.copy()
result2 = removeDuplicates_pythonic2(nums2)
print(f"Method 2 (List comprehension): {result2} unique elements")

# Method 3 (Two pointers)
nums3 = test_nums.copy()
result3 = removeDuplicates_pythonic3(nums3)
print(f"Method 3 (Two pointers): {result3} unique elements")


# class Solution:
#     def removeDuplicates(self, nums: List[int]) -> int:
#         a=0
#         unique=1
#         cm=1
#         n=len(nums)

#         while(cm<n):
#             if(nums[cm] != nums[cm-1]):
#                 a+=1
#                 nums[a]=nums[cm]
#                 unique+=1
#             cm+=1
#         return unique
        
#alternative methord